### PR TITLE
Add scale option to ToDtype. Remove ConvertDtype.

### PR DESCRIPTION
**Pull request by NicolasHug**
Closes https://github.com/pytorch/vision/issues/7756

this PR:

- adds a `scale` parameter to ToDtype, which only affects images or videos
- removes ConvetDtype to keep ConvertImageDtype. ConvertImageDtype now only supports images, not videos.
- removes all  dispatchers / kernels associated with `convert_.*`  and replace with `to_dtype*` dispatchers / kernels.
- When passing `ToDtype(torch.float32)` i.e. the `dtype` parameter is just a dtype, not a dict, we only convert images and videos - this is for BC with ConvertImageDtype and reduces adoption friction

(I may have removed too much stuff on the existing tests, we'll see with the CI)

### CODE REVIEW

File: torchvision/transforms/v2/_misc.py
Comment (NicolasHug) [inline torchvision/transforms/v2/_misc.py:246]: As discussed offline @pmeier , we need to decide on this

File: test/test_transforms_v2_refactored.py
Comment (NicolasHug) [inline test/test_transforms_v2_refactored.py:1746]: @pmeier

File: torchvision/transforms/v2/functional/_meta.py
Comment (pmeier) [inline torchvision/transforms/v2/functional/_meta.py:300]: Should this come after the dtype check? Functionally, this is a no-op here in case `image.dtype == dtype`, but it is still an unnecessary call.

```python
if image.dtype == dtype:
    return image
elif not scale:
    return image.to(dtype)
```

makes the behavior a little more clear and should be minimally more performant.

File: torchvision/transforms/v2/_misc.py
Comment (pmeier) [inline torchvision/transforms/v2/_misc.py:228]: Flagging so we don't forget.

File: torchvision/transforms/v2/_misc.py
Comment (pmeier) [reply torchvision/transforms/v2/_misc.py:246]: IMO, that is fine, since that is what most people likely want to do. However, should we also adapt `scale: Optional[bool] = None` in that case to mean: "if you only pass a single dtype, we are going to assume you want to convert and scale images and videos"? Basically having 

```python
if scale is None:
    scale = isinstance(dtype, torch.dtype)
self.scale = scale
```

in the constructor. That way `ToDtype` becomes a drop-in replacement for `ConvertImageDtype` with the ability to finetune.

File: torchvision/transforms/v2/functional/_meta.py
Comment (pmeier) [inline torchvision/transforms/v2/functional/_meta.py:369]:
```suggestion
    elif isinstance(inpt, datapoints._datapoint.Datapoint):
        return inpt.to(dtype)
    else:
        raise TypeError(
            f"Input can either be a plain tensor or an `Image` or `Video` datapoint,
            " f"but got {type(inpt)} instead.
        )
```

I prefer to sticking to regular dispatcher "layout" here:

1. It took me a few glances to actually see that we indeed have a `return inpt.to(dtype)` there. I think this comes from the fact that all other dispatchers don't do this. This is likely only going to happen to us maintainers since the code is clear, but still.
2. Right now, the dispatcher supports all tensors. Although I agree that `to_dtype` is general enough to allow this, all other dispatchers only support plain tensors and datapoints. And since for all other cases this is just a fancy wrapper around `t.to(dtype)`, I don't see a point of supporting that here.

File: torchvision/transforms/v2/_meta.py
Comment (pmeier) [inline torchvision/transforms/v2/_meta.py:67]: Non-blocking: We can probably drop the "strict" annotations here and align with all other transforms:

```suggestion
    def _transform(self, inpt: Any, params: Dict[str, Any]) -> Any:
```

File: torchvision/transforms/v2/functional/_meta.py
Comment (pmeier) [inline torchvision/transforms/v2/functional/_meta.py:350]: We still need to keep this alias for BC, right? Even more so, we need

```python
def convert_image_dtype(image: torch.Tensor, dtype: torch.dtype=torch.float32) -> torch.Tensor:
    return to_dtype_image_tensor(image, dtype=dtype, scale=True)
```

since I don't think JIT supports `functools.partial`.

File: torchvision/transforms/v2/_meta.py
Comment (pmeier) [inline torchvision/transforms/v2/_meta.py:69]: As written, this will convert plain tensor inputs into images. Should we just call the dispatcher and let it handle the (un-)wrapping?

```suggestion
        return F.to_dtype(inpt, dtype=self.dtype, scale=True)
```

Since we also set `_transformed_types` above, we'll only ever see plain tensors and images as `inpt`.

File: torchvision/transforms/v2/_misc.py
Comment (pmeier) [inline torchvision/transforms/v2/_misc.py:259]: Not sure if relevant in this PR: should we also do an `isinstance` check here afterwards? Might be useful in the context of custom datapoints. Same applies to the `fill` parameter on some other transforms. Maybe we can unify the functionality behind these "type-dict parameters"?

File: test/test_transforms_v2_refactored.py
Comment (pmeier) [inline test/test_transforms_v2_refactored.py:1745]: I'm guessing this should have been

```suggestion
        if input_dtype.is_floating_point and output_dtype is torch.uint8 and scale:
```

?

File: test/test_transforms_v2_refactored.py
Comment (pmeier) [reply test/test_transforms_v2_refactored.py:1746]: We don't have any tolerances set in the legacy v2 tests. However, we aren't testing full images there, but rather only

https://github.com/pytorch/vision/blob/cc0f9d02e1e7549e9e8ef1081a0d3b5d72f1297d/test/transforms_v2_kernel_infos.py#L1734-L1739

I'll have a look at the reference and get back to you.

File: test/test_transforms_v2_refactored.py
Comment (pmeier) [inline test/test_transforms_v2_refactored.py:1736]: Should we have at least one other integer dtype in there to cover this branch as well?

File: test/test_transforms_v2_refactored.py
Comment (pmeier) [inline test/test_transforms_v2_refactored.py:1739]: Nit: we have been calling it

```suggestion
    def test_image_correctness(self, input_dtype, output_dtype, device, scale):
```

File: test/test_transforms_v2_refactored.py
Comment (pmeier) [inline test/test_transforms_v2_refactored.py:1698]: Just for my understanding: we are using the catch-all key because we maybe go for the whole "single dtype means image" thing, right? Because otherwise, just passing the dtype should be sufficient, correct?

File: test/test_transforms_v2_refactored.py
Comment (pmeier) [reply test/test_transforms_v2_refactored.py:1683]: I don't understand the comment. I'll run it myself and see if I understand it there.

File: test/test_transforms_v2_refactored.py
Comment (pmeier) [inline test/test_transforms_v2_refactored.py:131]: Not against it, but are we expecting more kernels to set this to False?

File: test/test_transforms_v2_refactored.py
Comment (pmeier) [inline test/test_transforms_v2_refactored.py:73]: Good catch! Should we rename the parameter to `obj` or the like, since the input is no longer just a function?

File: test/test_transforms_v2_refactored.py
Comment (pmeier) [inline test/test_transforms_v2_refactored.py:1661]: Maybe

```suggestion
            expect_same_dtype=input_dtype is output_dtype,
```

**Review by pmeier — COMMENTED** [review]
I have a bunch of comments, but overall looks solid.

File: test/test_transforms_v2_refactored.py
Comment (pmeier) [reply test/test_transforms_v2_refactored.py:1746]: Ok, I can reproduce the issue when using proper images in the legacy tests. The error we are seeing stems from

https://github.com/pytorch/vision/blob/cc0f9d02e1e7549e9e8ef1081a0d3b5d72f1297d/torchvision/transforms/v2/functional/_meta.py#L321-L332

Two issues here:

1. Our reference is wrong.
   https://github.com/pytorch/vision/blob/cc0f9d02e1e7549e9e8ef1081a0d3b5d72f1297d/test/transforms_v2_kernel_infos.py#L1696
   It should be `round` rather than `int` following the comment above.
2. Albeit fewer, even with 1. fixed I'm still seeing some off-by-one errors. I've checked and this just stems from the fact that the more performant strategy we are using in the kernel, can produce slightly different results than the exact one we have as reference.

So an `atol=1` is fine in case we transform from float to int.

File: test/test_transforms_v2_refactored.py
Comment (pmeier) [reply test/test_transforms_v2_refactored.py:1683]: Ah, ok, you are saying that we need to disable this check, since there is no corresponding method on the datapoints and hence the test would fail, correct? If so, I'm ok to disable this here with the comment and re-visit after we have consensus on #7477.

File: test/test_transforms_v2_refactored.py
Comment (NicolasHug) [reply test/test_transforms_v2_refactored.py:131]: I don't think so (more precisely: I don't know). We still want the rest of the checks to be done for ToDtype though. Is there a better way than to add a parameter?

File: test/test_transforms_v2_refactored.py
Comment (NicolasHug) [reply test/test_transforms_v2_refactored.py:1683]: yes sorry this was pretty unclear. I'll make the comment easier to understand.

File: torchvision/transforms/v2/_misc.py
Comment (NicolasHug) [reply torchvision/transforms/v2/_misc.py:246]: > should we also adapt scale

This is the one thing where I think we should diverge from `ConvertImageDtype`: one of the most confusing thing about it is that it scales, while there's absolutely nothing in its name that makes it obvious. From https://github.com/pytorch/vision/issues/7756 

> It should be absolutely clear that ConvertDtype also scales the values

I'd rather enforce that users must pass `scale=True` if they want scaling to happen. Any concern?

File: torchvision/transforms/v2/_misc.py
Comment (NicolasHug) [reply torchvision/transforms/v2/_misc.py:259]: You mean ensuring that the resulting `dtype` is an actual dtype (like torch.float32)? I assume that users would get a decent error message down the road anyway, at least when `.to()` gets called?

File: torchvision/transforms/v2/functional/_meta.py
Comment (NicolasHug) [reply torchvision/transforms/v2/functional/_meta.py:300]: I'll do it but TBH  I don't really follow the reasoning behind the change

Comment (NicolasHug) [reply torchvision/transforms/v2/functional/_meta.py:300]: The `elif` part actually adds confusion to me because it suggests these 2 blocks are related when in reality they're really not

File: torchvision/transforms/v2/functional/_meta.py
Comment (NicolasHug) [reply torchvision/transforms/v2/functional/_meta.py:350]: ah yeah, I didn't realize we had this in V1. Thanks

File: torchvision/transforms/v2/_misc.py
Comment (NicolasHug) [reply torchvision/transforms/v2/_misc.py]: I implemented what we agree on in [00aef9b](https://github.com/pytorch/vision/pull/7759/commits/00aef9b038f723df474a56152094543231fe615c) (leaving-out the `scale` thing for now)

File: torchvision/transforms/v2/_misc.py
Comment (vfdev-5) [inline torchvision/transforms/v2/_misc.py:244]: Removing this behaviour we would like to go away from `defaultdict(lambda: dtype)` ?

Comment (NicolasHug) [reply torchvision/transforms/v2/_misc.py:244]: Yes, this is one of the goals https://github.com/pytorch/vision/issues/7756#issuecomment-1649383342

Comment (vfdev-5) [reply torchvision/transforms/v2/_misc.py:244]: But should not this be done in a separate PR and everywhere it is used ?

Comment (NicolasHug) [reply torchvision/transforms/v2/_misc.py:244]: If this gets accepted here I was going to submit a follow-up PR to do the same changes for the `fill` parameters. Are there other places where this pattern is used?

Comment (vfdev-5) [reply torchvision/transforms/v2/_misc.py:244]: OK, sounds good to me.

PermuteDimensions, TransposeDimensions and `_setup_fill_arg`

File: torchvision/transforms/v2/functional/_meta.py
Comment (pmeier) [reply torchvision/transforms/v2/functional/_meta.py]: Re-reading the original implementation: I was wrong. Since we already `return`'ed in both branches, this should have no effect on the performance. Thus, feel free to revert to what you had.

Still, I would put the dtype check at the top, because that is the "more important" one. Again, just style / personal preference. Ignore if you feel different.



> The `elif` part actually adds confusion to me because it suggests these 2 blocks are related when in reality they're really not

Fair enough. Happy with a second `if` as well.

File: test/test_transforms_v2_refactored.py
Comment (NicolasHug) [inline test/test_transforms_v2_refactored.py:1763]: I added this because when the dtype parameters isn't a dict and when the input is a single bbox or a single mask (as is the case here), the input just gets passed-through. I'm converting the dtype into a dict so that we also test the rest of the code-paths for those single bboxes and masks. Not incredibly critical, just for coverage.

**Review by pmeier — APPROVED** [review]
Some comments, but nothing blocking. LGTM if CI is green. Thanks Nicolas!

File: test/test_transforms_v2_refactored.py
Comment (NicolasHug) [reply test/test_transforms_v2_refactored.py:133]: I have a slight preference for adding the parameter, because otherwise we'd have to change both implementation of `check_kernel` if we ever needed to update it.
(looks like I'm advocating for this single entry point after all :p )

File: test/test_transforms_v2_refactored.py
Comment (NicolasHug) [reply test/test_transforms_v2_refactored.py:1838]: We do (from the parametrization), although I kept `img` as a name which I agree is confusing. I'll use `inpt` instead

File: test/test_transforms_v2_refactored.py
Comment (NicolasHug) [reply test/test_transforms_v2_refactored.py:1861]: They are redundant but I slightly prefer keeping both, because just reading

```
assert out["mask"].dtype == torch.int64
```

doesn't make it obvious that the output dtype is different from the input dtype

File: test/test_transforms_v2_refactored.py
Comment (pmeier) [reply test/test_transforms_v2_refactored.py:1830]: Thanks, I missed that.

**Comment by NicolasHug** [comment]
Tests are green in [92f2588](https://github.com/pytorch/vision/pull/7759/commits/92f2588852b105d209a2e943a6e742446831f85d) except for the linter which I just fixed. Merging, thanks for the reviews!

**Pull request merged**
